### PR TITLE
Server API Now Has Health Check Option

### PR DESCRIPTION
**Pull request by kevinkreiser**
There's now a cheap health check option in the server API. You need to provide it a health check pattern matcher functor to tell when a request is a health check request and a canned reponse to that request. This allows you to configure prime_server to work with a couple different existing monitoring systems like aws which has a health checking  endpoint built into the elbs or in kubernetes which have the liveness checks.

I've also added graceful shutdown mechanics that should play nice with runners like kubernetes and potentially other runners that use SIGTERM to ask the application to shutdown gracefully (common practice). This works by configuring two grace periods in the API. The **draining period** (in which the server should handle the remaining outstanding requests or synchronize the program state to disk or clean up any resources, etc) and the **shutdown period** in which the underlying threads should terminate. we achieve it by blocking the SIGTERM signal in the main and all its children threads, we then spawn a daemon thread that just sits and waits for SIGTERM forever. Once the signal is caught by that thread, the thread uses the configured seconds above to allow the other threads to clean up and shutdown.

### CODE REVIEW

File: .clang-format [hunk 1]
```diff
@@ -17,7 +17,7 @@ AllowShortLoopsOnASingleLine: false
 AlwaysBreakAfterDefinitionReturnType: None
 AlwaysBreakAfterReturnType: None
 AlwaysBreakBeforeMultilineStrings: false
-AlwaysBreakTemplateDeclarations: false
+AlwaysBreakTemplateDeclarations: Yes
```
Comment (kevinkreiser) [inline .clang-format:20]: i hate templates on the same line so some of the diffs are from this

File: prime_server/prime_server.hpp [hunk 2]
```diff
@@ -70,7 +73,9 @@ template <class request_container_t, class request_info_t> class server_t {
            const std::string& interrupt_endpoint,
            bool log = false,
            size_t max_request_size = DEFAULT_MAX_REQUEST_SIZE,
-           uint32_t request_timeout = DEFAULT_REQUEST_TIMEOUT);
+           uint32_t request_timeout = DEFAULT_REQUEST_TIMEOUT,
+           const std::function<bool(const request_container_t&)>& health_check_matcher = {},
+           const std::string& health_check_response = {});
```
Comment (kevinkreiser) [inline prime_server/prime_server.hpp:78]: these are the new options as describted, a matcher to identify when a request is a health check and the response to that when we see one

File: prime_server/zmq_helpers.hpp [hunk 3]
```diff
@@ -25,11 +25,10 @@ struct message_t {
...
       void* data,
       size_t size,
-      void (*free_function)(void*, void*) = [](void* ptr, void* /*hint*/) {
-        delete[] static_cast<unsigned char*>(ptr);
+      void (*free_function)(void*, void*) = [](void* p, void*) {
+        delete[] static_cast<unsigned char*>(p);
       });
-  explicit message_t(size_t size = 0);
-  void reset(size_t size = 0);
+  explicit message_t(size_t size = 0, const void* data = nullptr);
```
Comment (kevinkreiser) [inline prime_server/zmq_helpers.hpp:31]: this is the only material change. we had this default constructor here just for testing but it turned out to be nice to have a long lived zmq message lying aroudn to reply with that actually had data in it.

File: src/prime_server.cpp [hunk 4]
```diff
@@ -263,18 +267,28 @@ bool server_t<request_container_t, request_info_t>::enqueue(const zmq::message_t
...
       parsed_request.log(info.id);
+
     // remember we are working on it
     request.enqueued.emplace_back(static_cast<typename decltype(requests)::key_type>(info));
     requests.emplace(request.enqueued.back(), requester);
     request_history.emplace_back(std::move(info));
+
+    // if it was a health check we reply immediately
+    if (health_check)
+      dequeue(request_history.back(), health_check_response);
```
Comment (kevinkreiser) [inline src/prime_server.cpp:291]: this is the actual change to enable the functionality. i didnt go crazy here. if you opted into the health check functionality and we see a request is a health check, we dont forward it to the proxy/load balancer, instead we enqueue as if we did but then when immediately dequeue it (which is what replies to the client) with the canned response. i could have avoided enquing and dequing it and just replied directly but liked that the logging and also the connection closing logic is encapsulated in those functions so i kept it simple at the sake of a tiny bit of performance degradation

File: src/zmq_helpers.cpp [hunk 5]
```diff
@@ -42,11 +42,14 @@ message_t::message_t(void* data, size_t size, void (*free_function)(void*, void*
...
 }
-message_t::message_t(size_t size) {
+message_t::message_t(size_t size, const void* data) {
   // make the c message
   zmq_msg_t* message = new zmq_msg_t();
   if (zmq_msg_init_size(message, size) != 0)
     throw std::runtime_error(zmq_strerror(zmq_errno()));
+  // copy the data into it
+  if (size && data)
+    memcpy(zmq_msg_data(message), data, size);
```
Comment (kevinkreiser) [inline src/zmq_helpers.cpp:52]: yeah so before this constructor didnt do anything it was like the default make a message that is empty or random junk initialized but of a certain size. now you can actually pass data to it to manage

File: src/zmq_helpers.cpp [hunk 6]
```diff
@@ -55,13 +58,6 @@ message_t::message_t(size_t size) {
     delete message;
   });
 }
-void message_t::reset(size_t size) {
```
Comment (kevinkreiser) [inline src/zmq_helpers.cpp:58]: this function was unused so i yanked it

File: test/http.cpp [hunk 7]
```diff
@@ -415,6 +418,30 @@ void test_large_request() {
   client.batch();
 }
 
+void test_health_check() {
```
Comment (kevinkreiser) [inline test/http.cpp:421]: the changes in the tests are straight forward. use the new options and test that they work

File: README.md [hunk 8]
```diff
@@ -246,6 +246,9 @@ We basically just have to define a `work` function/object/lambda whose signature
 
 ```c++
 int main(void) {
+  //when we hear SIGTERM, wait 28 seconds to drain traffic then wait a second for the threads to stop
```
[the remark this answers is not in the document]
Comment (kevinkreiser) [reply README.md:249]: great question! its because a lot of systems use 30 seconds as the default, so to avoid SIGKILL we want to make sure we are done with everything before 30 seconds, so i wanted to have 29 seconds in total, drain(28) + shutdown(1) == 29 < 30

[the remark this answers is not in the document]
Comment (mandeepsandhu) [reply README.md:249]: I might've missed it, but does `prime_server` stop accepting new connections once it goes into the `draining` state?

File: src/prime_server.cpp [hunk 9]
```diff
@@ -15,6 +21,46 @@ struct interrupt_t : public std::runtime_error {
...
+        shutting_down = true;
+        sleep(shutdown_seconds);
+        exit(0);
+      } else {
+        logging::ERROR("Could not wait for SIGTERM, graceful shutdown disabled");
+      }
+    }).detach();
+  }
+
+  bool draining;
```
Comment (mandeepsandhu) [inline src/prime_server.cpp:60]: should these be made an `atomic_bool`s? I guess the downside is 1 extra loop of the main loop of prime server (or could there be more subtle caching issues on a multi-core CPU? maybe I'm overthinking this :smile: ).

Comment (kevinkreiser) [reply src/prime_server.cpp:60]: i considered it but didnt think it was needed because:

1. only the signal handler writes to while other threads are reading it
2. writes to primatives are atomic on modern architectures

Here's a quick explanation of what I'm relying on: https://stackoverflow.com/questions/52424277/c-is-the-assignment-of-a-value-to-a-primitive-data-type-e-g-bool-an-atomic

I'll benchmark it with atomic and see if it matters at all (hopefully not)

Comment (mandeepsandhu) [reply src/prime_server.cpp:60]: yeah you're right, there's no danger of partial writes. I was more thinking about different threads on different cores not seeing the writes to the bools because of caching (the cache coherency issue that the SO thread alluded to as well).

Comment (kevinkreiser) [reply src/prime_server.cpp]: atomic was fine so i added it. going to merge. ive tested this extensively with the branch i have going in valhalla. preparing this for release

Comment (mandeepsandhu) [reply src/prime_server.cpp]: Thanks for adding this feature @kevinkreiser :pray: !
